File: examples/paper_example04.jl
using OptimalTransportNetworks
using Random
using Plots

# Initialize parameters
param = init_parameters(K = 100, rho = 0, labor_mobility = false)
width, height = 13, 13

# Create graph
g0 = create_graph(param, width, height, type = "map")

# Set fundamentals
g0[:Zjn] = ones(g0[:J], 1) .* 1e-7  # matrix of productivity (not 0 to avoid numerical glitches)
g0[:Lj] = ones(g0[:J]) .* 1e-7  # matrix of population

Ni = find_node(g0, ceil(width/2), ceil(height/2)) # center
g0[:Zjn][Ni] = 1 # more productive node
g0[:Lj][Ni] = 1 # more productive node

Random.seed!(10) # use 5, 8, 10 or 11
nb_cities = 20 # draw a number of random cities in space
for i in 1:nb_cities-1
    newdraw = false
    while newdraw == false
        j = round(Int, 1 + rand() * (g0[:J] - 1))
        if g0[:Lj][j] <= 1e-7
            newdraw = true
            g0[:Lj][j] = 1
        end
    end
end

g0[:hj] = g0[:Hj] ./ g0[:Lj]
g0[:hj][g0[:Lj] .== 1e-7] .= 1  # catch errors in places with infinite housing per capita

# Draw geography
z = zeros(g0[:J]) # altitude of each node
geographies = Dict()
geographies[:blank] = (z = z, obstacles = nothing)

g = apply_geography(g0, geographies[:blank])

# Blank geography
results = Dict(:blank => optimal_network(param, g))

# Mountain
mountain_size = 0.75
mountain_height = 1
mount_x = 10
mount_y = 10
geographies[:mountain] = (z = mountain_height * exp.(-((g0[:x] .- mount_x).^2 .+ (g0[:y] .- mount_y).^2) / (2 * mountain_size^2)),
                          obstacles = nothing)
g = apply_geography(g0, geographies[:mountain], alpha_up_i = 10, alpha_down_i = 10)
results[:mountain] = optimal_network(param, g)

# Adding river and access by land
geographies[:river] = (z = geographies[:mountain].z, 
                       obstacles = [6 + (1-1)*width 6 + (2-1)*width;
                          6 + (2-1)*width 6 + (3-1)*width;
                          6 + (3-1)*width 7 + (4-1)*width;
                          7 + (4-1)*width 8 + (5-1)*width;
                          8 + (5-1)*width 9 + (5-1)*width;
                          11 + (5-1)*width 12 + (5-1)*width;
                          12 + (5-1)*width 13 + (5-1)*width])

g = apply_geography(g0, geographies[:river], alpha_up_i = 10, alpha_down_i = 10)
results[:river] = optimal_network(param, g)

# Reinit and put another river and bridges
geographies[:bridges] = (z = mountain_height * exp.(-((g0[:x] .- mount_x).^2 .+ (g0[:y] .- mount_y).^2) / (2 * mountain_size^2)),
                         obstacles = [6 + (1-1)*width 6 + (2-1)*width;
                             6 + (2-1)*width 6 + (3-1)*width;
                             6 + (3-1)*width 7 + (4-1)*width;
                             7 + (4-1)*width 8 + (5-1)*width;
                             8 + (5-1)*width 9 + (5-1)*width;
                             9 + (5-1)*width 10 + (5-1)*width;
                             10 + (5-1)*width 11 + (5-1)*width;
                             11 + (5-1)*width 12 + (5-1)*width;
                             12 + (5-1)*width 13 + (5-1)*width])

g = apply_geography(g0, geographies[:bridges], alpha_up_i = 10, alpha_down_i = 10, 
                    across_obstacle_delta_i = 2)
results[:bridges] = optimal_network(param, g)

# Allowing for water transport
geographies[:water_transport] = (z = mountain_height * exp.(-((g0[:x] .- mount_x).^2 .+ (g0[:y] .- mount_y).^2) / (2 * mountain_size^2)),
                                 obstacles = [6 + (1-1)*width 6 + (2-1)*width;
                                              6 + (2-1)*width 6 + (3-1)*width;
                                              6 + (3-1)*width 7 + (4-1)*width;
                                              7 + (4-1)*width 8 + (5-1)*width;
                                              8 + (5-1)*width 9 + (5-1)*width;
                                              9 + (5-1)*width 10 + (5-1)*width;
                                              10 + (5-1)*width 11 + (5-1)*width;
                                              11 + (5-1)*width 12 + (5-1)*width;
                                              12 + (5-1)*width 13 + (5-1)*width])

g = apply_geography(g0, geographies[:water_transport], alpha_up_i = 10, alpha_down_i = 10, 
                    across_abstacle_delta_i = 2, 
                    along_obstacle_delta_i = 0.5,
                    along_abstacle_delta_tau = 1)
results[:water_transport] = optimal_network(param, g)

# Increasing returns to transport
param[:gamma] = 2
geographies[:increasing_returns] = geographies[:water_transport]
results[:increasing_returns] = optimal_network(param, g)

# Plot results
simulation = ["blank", "mountain", "river", "bridges", "water_transport", "increasing_returns"]
obstacles = ["off", "off", "on", "on", "on", "on"]
plots = Vector{Any}(undef, length(simulation)) 

i = 0
for s in simulation
    i += 1
    plots[i] = plot_graph(g0, results[Symbol(s)][:Ijk], 
                          geography = geographies[Symbol(s)], obstacles = obstacles[i] == "on",
                          mesh = true, mesh_transparency = 0.2,
                          node_sizes = results[Symbol(s)][:cj], 
                          node_shades = g0[:Zjn], node_color = :seismic,
                          edge_min_thickness = 1, edge_max_thickness = 4)
    title!(plots[i], "Geography $(s)")
end

# Combine plots
final_plot = plot(plots..., layout = (2, 3), size = (3*400, 2*400))
display(final_plot)

File: OptimalTransportNetworkToolbox_v.1.0.4b/Code/paper_example04.m
%{
==============================================================
 OPTIMAL TRANSPORT NETWORKS IN SPATIAL EQUILIBRIUM
 by P. Fajgelbaum, E. Schaal, D. Henricot, C. Mantovani 2017-19
 ================================================ version 1.0.4

paper_example04.m: script that generate illustrative example 04 from the
paper.

-----------------------------------------------------------------------------------
REFERENCE: "Optimal Transport Networks in Spatial Equilibrium" (2019) by Pablo D.
Fajgelbaum and Edouard Schaal.

Copyright (c) 2017-2018, Pablo D. Fajgelbaum, Edouard Schaal
pfajgelbaum@ucla.edu, eschaal@crei.cat

This code is distributed under BSD-3 License. See LICENSE.txt for more information.
-----------------------------------------------------------------------------------
%}

clear;
close all;

%% ------------------------------------
% 4.3.1 Geography
% -------------------------------------

width=13; height=13;

% Init and Solve network

param = init_parameters('TolKappa',1e-5,'K',100,'LaborMobility','off');
[param,g]=create_graph(param,width,height,'Type','map'); 

% set fundamentals

rng(5);
param.N = 1;
param.Zjn = param.minpop*ones(g.J,1); % matrix of productivity
param.Hj = ones(g.J,1); % matrix of housing
param.Lj = 0*ones(g.J,1); % matrix of population

Ni=find_node(g,5,5); % center
param.Zjn(Ni)=1; % more productive node
param.Lj(Ni)=1; % more productive node

nb_cities=20; % draw a number of random cities in space
for i=1:nb_cities-1
    newdraw=false;
    while newdraw==false
        j=round(1+rand()*(g.J-1));
        if param.Lj(j)<=param.minpop
            newdraw=true;
            param.Lj(j)=1;
        end
    end
end

param.hj=param.Hj./param.Lj;
param.hj(param.Lj==0)=1;

% --------------
% Draw geography

z=zeros(g.J,1); % altitude of each node
obstacles = [];

geography = struct('z',z,'obstacles',obstacles);
g = apply_geography( g, geography );

param0=param; % store initial params
g0=g; % store initial graph


%% Blank geography
results(1)=optimal_network(param,g);

%% Mountain
mountain_size=.75;
mountain_height=1;
mount_x=10;
mount_y=10;
geography(2)=geography(1);
geography(2).z=mountain_height*exp(-((g.x-mount_x).^2 + (g.y-mount_y).^2)/(2*mountain_size^2));

g = apply_geography( g, geography(2),'AlphaUp_i',10,'AlphaDown_i',10 );
results(2)=optimal_network(param,g);

%% Adding river and access by land
g=g0;
geography(3)=geography(2);
geography(3).obstacles = [6 + (1-1)*width, 6 + (2-1)*width;
    6 + (2-1)*width, 6 + (3-1)*width;
    6 + (3-1)*width, 7 + (4-1)*width;
    7 + (4-1)*width, 8 + (5-1)*width;
    8 + (5-1)*width, 9 + (5-1)*width;
    11 + (5-1)*width, 12 + (5-1)*width;
    12 + (5-1)*width, 13 + (5-1)*width]; % Nobj x 2 matrix of (i,j) pairs of locations where a geographic barrier should be drawn

g = apply_geography( g, geography(3),'AcrossObstacleDelta_i', inf, 'AlphaUp_i',10,'AlphaDown_i',10 );
results(3)=optimal_network(param,g);

%% Reinit and put another river and bridges
g=g0;
geography(4)=geography(1);
geography(4).z=mountain_height*exp(-((g.x-mount_x).^2 + (g.y-mount_y).^2)/(2*mountain_size^2));

geography(4).obstacles = [6 + (1-1)*width, 6 + (2-1)*width;
    6 + (2-1)*width, 6 + (3-1)*width;
    6 + (3-1)*width, 7 + (4-1)*width;
    7 + (4-1)*width, 8 + (5-1)*width;
    8 + (5-1)*width, 9 + (5-1)*width;
    9 + (5-1)*width, 10 + (5-1)*width;
    10 + (5-1)*width, 11 + (5-1)*width;
    11 + (5-1)*width, 12 + (5-1)*width;
    12 + (5-1)*width, 13 + (5-1)*width];

g = apply_geography( g, geography(4), 'AlphaUp_i',10,'AlphaDown_i',10, 'AcrossObstacleDelta_i',2,'AlongObstacleDelta_i',inf );
results(4)=optimal_network(param,g);

%% Allowing for water transport
g=g0;
geography(5)=geography(1);
geography(5).z=mountain_height*exp(-((g.x-mount_x).^2 + (g.y-mount_y).^2)/(2*mountain_size^2));
geography(5).obstacles = [6 + (1-1)*width, 6 + (2-1)*width;
    6 + (2-1)*width, 6 + (3-1)*width;
    6 + (3-1)*width, 7 + (4-1)*width;
    7 + (4-1)*width, 8 + (5-1)*width;
    8 + (5-1)*width, 9 + (5-1)*width;
    9 + (5-1)*width, 10 + (5-1)*width;
    10 + (5-1)*width, 11 + (5-1)*width;
    11 + (5-1)*width, 12 + (5-1)*width;
    12 + (5-1)*width, 13 + (5-1)*width];

g = apply_geography( g, geography(5), 'AlphaUp_i',10,'AlphaDown_i',10, 'AcrossObstacleDelta_i',2,'AlongObstacleDelta_i',.5 );
results(5)=optimal_network(param,g);

%% Increasing returns to transport
param.gamma=2;
geography(6)=geography(5);
results(6)=optimal_network(param,g);

%% Plot results
close all;

s={'geography_blank','geography_mountain','geography_river','geography_bridges','geography_water_transport','geography_increasing_returns'};
obstacles = {'off','off','on','on','on','on'};

for j=1:length(s)
    fig=figure('Units','inches','Position',[0,0,7.5,5],'Name',char(s(j)));

    % Plot network
    plot_graph( param,g,results(j).Ijk,'Geography','on','GeographyStruct',geography(j),'Sizes',3*results(j).cj.*(param.Lj>param.minpop)/max(results(j).cj),'Mesh','on','MeshTransparency',.2,'Obstacles',char(obstacles(j)), 'Shades',param.Zjn, 'MinEdgeThickness', 1.5 );

    % Save
%     print('-depsc2',[graph_path,char(s(j)),'.eps']);
%     saveas(fig,[graph_path,char(s(j)),'.png']);
end
